torch/_tensor.py

```diff
@@ -241,8 +241,10 @@ def backward(self, gradient=None, retain_graph=None, create_graph=False, inputs=
             inputs (sequence of Tensor): Inputs w.r.t. which the gradient will be
                 accumulated into ``.grad``. All other Tensors will be ignored. If not
                 provided, the gradient is accumulated into all the leaf Tensors that were
-                used to compute the attr::tensors. All the provided inputs must be leaf
-                Tensors.
+                used to compute the attr::tensors.
+                For a non-leaf provided input, its grad_fn will be evaluated and
```

Comment:
I am not convinced by this. It will most likely be cryptic for users.

What about moving it outside of the arg documentation and into a special `.. note::` block.
When `inputs` are provided and a given input is not a leaf, the current implementation will call its grad_fn (even though it is not strictly needed to get this gradients). It is an implementation detail on which the user should not rely. See https://github.com/pytorch/pytorch/pull/60521#issuecomment-867061780 for more details.

Reply:
Addressed in the latest PR.